Work completed so far to support v1p2

andrew-gardener opened the pull request:
- switched fixtures to v1p2
- removed profile extension contexts
- added profile property to events (and updated tests)
- added `Collectable` interface to finish v1p2 changes needed for Collection inheritance 
- added `storageName` to DigitalResource
- added SystemIdentifier and `otherIdentifiers` to Entity
- added `client` to Session and `host`, `ipAddress`, and `userAgent` to SoftwareApplication
- added `makeAnonymous` static helper function (to create help anonymous entities)

andrew-gardener commented:
Note that a lot of the fixtures changed (some new ones and a few renamed or removed). I haven't worked on adding any of the new fixtures yet (though I took care of the renamed/removed ones that cropped up)

andrew-gardener commented:
Going to work on adding missing fixture tests now

andrew-gardener commented:
finished adding the missing fixtures. they're all there expect `caliperEndpointConfigPayload.json` since it is for a v1p2 extension.

This should cover everything new for v1p2 or at least what's listed in the google doc (didn't go over the spec extensively to see if there is anything missed).

There is also the new `commonErrorFixtures` to handle but that can be in another PR

lsloan commented:
Hi, @andrew-gardener.  Got your review request.  I'll start reviewing ASAP.

How much of the v1p2 requirements does this fulfill?  Is there more that needs work?  I've recently finished working on one UMich project and have joined the work on MyLA (you're familiar with that), but I think I'll have some spare time to work on outstanding v1p2 requirements.

andrew-gardener commented:
@lsloan It should cover all the spec changes. I believe that the only work remaining should be adding fixtures tests for expected failures. Other than that, there are a few admin tasks on the spreadsheet (setup Travis-CI tests+builds and release new version).

Also just wanted to point out there is one place where I did something different than pure spec. [SystemIdentifier](https://github.com/IMSGlobal/caliper-php/blob/fa595784e7ed1015cad2341ac6e5e1819e2cd443/src/entities/SystemIdentifier.php#L18) extends entity and just unsets id (and uses two params, identifier and identifierType, for initialization). I mostly did it this way so it would be easier to implement but in the spec it is not based on Entity at all. I can change this if you want.

lsloan commented on src/events/Event.php:
```diff
@@ -140,6 +144,20 @@ public function setActor(entities\foaf\Agent $actor) {
+        $this->profile = $profile;
+        return $this;
+    }
+
```
Should allow setting null and types should indicate that it will accept/return null.

andrew-gardener replied:
fixed

lsloan commented:
> @lsloan It should cover all the spec changes. I believe that the only work remaining should be adding fixtures tests for expected failures. Other than that, there are a few admin tasks on the spreadsheet (setup Travis-CI tests+builds and release new version).
> 
> Also just wanted to point out there is one place where I did something different than pure spec. [SystemIdentifier](https://github.com/IMSGlobal/caliper-php/blob/fa595784e7ed1015cad2341ac6e5e1819e2cd443/src/entities/SystemIdentifier.php#L18) extends entity and just unsets id (and uses two params, identifier and identifierType, for initialization). I mostly did it this way so it would be easier to implement but in the spec it is not based on Entity at all. I can change this if you want.

Yeah, I saw that.  So the spec. says that object doesn't need an ID?

I think using Entity this way is hackish, but not necessarily bad.  If it serializes to the correct JSON as given in the fixtures, it's probably OK.

Do you think anyone will ever need to do type comparisons of SystemIdentifier where it **_shouldn't_** evaluate as an extension of Entity?

If it wasn't an extension of Entity, a lot of code from Entity would need to be copied into this class, right?  (Or some generalization would need to be done.)

andrew-gardener commented on src/profiles/Profile.php:
```diff
@@ -0,0 +1,23 @@
+        SURVEY = 'SurveyProfile',
+        TOOL_LAUNCH = 'ToolLaunchProfile',
+        TOOL_USE = 'ToolUseProfile',
+        VIEWED = 'Viewed';
```
oh it shouldn't actually be there (might have been a copy+paste error) as there isn't a ViewedProfile

andrew-gardener commented on src/events/SessionEvent.php:
```diff
@@ -2,6 +2,7 @@
 
 namespace IMSGlobal\Caliper\events;
 
+use IMSGlobal\Caliper\profiles\Profile;
```
removed

andrew-gardener commented on src/entities/Entity.php:
```diff
@@ -135,6 +138,33 @@ public function setName($name) {
+    }
+
+    /**
+     * @param SystemIdentifier[] $otherIdentifiers
```
fixed

andrew-gardener commented on src/events/NavigationEvent.php:
```diff
@@ -31,12 +36,31 @@ public function getObject() {
             $this->object = $object;
-            if ($object instanceof Questionnaire || $object instanceof QuestionnaireItem) {
-                $this->setContext(new Context(Context::SURVEY_PROFILE_EXTENSION));
+            if ($this->profile == profiles\Profile::SURVEY) {
```
changed to `===`

https://raw.githack.com/IMSGlobal/caliper-spec/develop/caliper-spec-v1p2.html#profile-survey NavigationEvent and ViewedEvent normally constrains it to DigitalResource but the Survey profile constrains it further to just Questionnaire and QuestionnaireItem

andrew-gardener commented on src/events/NavigationEvent.php:
```diff
@@ -31,12 +36,31 @@ public function getObject() {
+    }
+
+    /**
+     * @param entities\DigitalResource $target
```
fixed and modified Event to be like this as well

andrew-gardener commented on src/events/GradeEvent.php:
```diff
@@ -2,6 +2,7 @@
 
 namespace IMSGlobal\Caliper\events;
 
+use IMSGlobal\Caliper\profiles;
```
removed

lsloan commented:
@andrew-gardener, I'll keep watching this space for messages from you, but also remember to click the "re-request review" circular arrows beside my name after you're done with revisions based on my comments.

lsloan commented:
I'm sorry to be a PITA, but I'm starting a second review.  I checked out your branch and ran the tests.  They've failed because of a couple of errors so far.  Just in case there are more, I thought marking them all in a review would be the best way to show you where I found errors.

andrew-gardener commented:
@lsloan no problem about doing another review. I'm running PHP 7.4.1 locally and haven't tested against any other versions when doing these changes (this is one good thing about when Travis-CI would be in place since it will be easy to automatically test different environments/settings).

Anyways, I've fixed all the issues from the first run of the code review and also changed SystemIdentifier to be more correct (and less hacky) though there is a bit more code duplication now.

lsloan commented on tests/EntitySystemIdentifierTest.php:
```diff
@@ -0,0 +1,22 @@
+
+        $this->setTestObject(
+            (new SystemIdentifier('https://example.edu/users/554433', new SystemIdentifierType(SystemIdentifierType::LTI_USERID)))
+                ->setSource( (new SoftwareApplication('https://example.edu')) ),
```
Error from trailing comma.

andrew-gardener replied:
fixed

lsloan commented on tests/EventNavigationNavigatedToCollectionItemTest.php:
```diff
@@ -1,62 +1,70 @@
+                        ->setName('Video Collection')
+                        ->setKeywords(['collection', 'videos'])
+                        ->setDateCreated(new \DateTime('2016-08-01T06:00:00.000Z'))
+                        ->setDateModified(new \DateTime('2016-09-02T11:30:00.000Z')),
```
Error from trailing comma.

andrew-gardener replied:
fixed

lsloan commented on tests/EventNavigationNavigatedToCollectionItemTest.php:
```diff
@@ -1,62 +1,70 @@
+use IMSGlobal\Caliper\events\NavigationEvent;
+use IMSGlobal\Caliper\entities\DigitalResourceCollection;
+use IMSGlobal\Caliper\entities\Entity;
+use IMSGlobal\Caliper\entities\media\VideoObject;
```
While looking for the error on line 44, my IDE warned me that some of these imported modules are not used.

Check and clean up these before we release.

andrew-gardener replied:
got an extension for my IDE to point these out and went through everything removing them

lsloan commented:
@andrew-gardener, shall I merge this PR?

andrew-gardener commented:
Yes please

lsloan commented:
Done!
